packages/compiler/src/server/serverlib.ts

```diff
@@ -339,6 +343,112 @@ export function createServer(host: ServerHost): Server {
     }
   }
 
+  async function updateImportsOnFileChange(changes: FileEvent[]): Promise<void> {
+    const tspChanges = changes.filter((change) => change.uri.endsWith(".tsp"));
+    const filePathChangedMap = new Map<string, string>();
+
+    if (tspChanges.length === 0 || tspChanges.length % 2 !== 0) {
+      return;
+    }
+
+    // type 1: created, type 3: deleted
+    // Modify file name or move the file to a new location,
+    // the original file is marked as delete, and the new file is marked as Create
+    if (tspChanges.length === 2) {
+      // Only one file is renamed or moved
+      const oldFilePath = tspChanges.find((c) => c.type === 3);
+      const newFilePath = tspChanges.find((c) => c.type === 1);
+      if (oldFilePath && newFilePath) {
+        filePathChangedMap.set(
+          await fileService.getPath({ uri: oldFilePath.uri }),
+          await fileService.getPath({ uri: newFilePath.uri }),
+        );
+      }
+    } else {
+      for (const change of tspChanges) {
+        if (change.type === 1) {
+          const newFilePath = await fileService.getPath({ uri: change.uri });
+          const fileName = getBaseFileName(newFilePath);
+          const oldChange = tspChanges.find((c) => c.type === 3 && c.uri.endsWith(fileName));
+          if (oldChange) {
+            const oldFilePath = await fileService.getPath({ uri: oldChange.uri });
+            filePathChangedMap.set(oldFilePath, newFilePath);
+          }
+        }
+      }
+    }
+
+    if (filePathChangedMap.size !== tspChanges.length / 2) {
+      log({
+        level: "debug",
+        message:
+          "The number of files marked for deletion(type=3) and creation(type=1) is not equal, " +
+          "they must appear in pairs, see the input parameters for details: \n" +
```

Comment:
must is too strong and maybe confusing. It's still possible to be a valid scenario not for rename. I think you can just mention can't pair files properly and skip the rename handling is enough.